### PR TITLE
feat: oauth2 to generate kubeconfig

### PR DESCRIPTION
**Pull request by Abingcbc**
https://github.com/labring/sealos/issues/1173

## Oauth2 Login and generate kubeconfig for user

https://user-images.githubusercontent.com/34068205/178107579-8a99f226-d79f-4b87-80f9-d27b5abe70c7.mov

## Kubeconfig validation

https://user-images.githubusercontent.com/34068205/176468684-915be394-7f18-4c09-a4a9-d6702ecc330f.mov

TODO:

- [x] Provide casdoor kubernetes yaml (all in one).
- [x] Using cloud.sealos.io as the login home page, change login logo.
- [ ] Persist Casdoor data based on openebs
- [ ] Blog or document about how to integrate casdoor and sealos.

### Start up
Sealos Desktop Auth and Casdoor both run on sealos. After they started, Sealos Desktop Auth will load the configuration file like the format of `conf/desktop.yaml.template`. Then it will start to init Casdoor.

1. Get deployment of Casdoor
2. Generate Casdoor init-data file `init_data.json`
3. Create a ConfigMap from this init-data file
4. Mount the new init-data file on Casdoor pod by updating the pod template in deployment
5. K8s will automatically restart the Casdoor pod

### Generate KubeConfig

1. Create `ServiceAccount`
2. Create `ClusterRoleBinding`
3. Create a [Scret of token](https://kubernetes.io/docs/concepts/configuration/secret/#service-account-token-secrets) for `ServiceAccount`. In k8s v1.22 or before, k8s will automatically create this secret. However, after v1.22, k8s will not. 
Therefore, we should manually create an empty secret that is annotated with `kubernetes.io/service-account.name`. Then k8s will automatically fill the token in this Secret.
4. Wait for k8s's fill and get the token
5. Generate kubeconfig which uses the token as authentication

### CODE REVIEW

**Comment by Abingcbc** [comment]
@fanux @cuisongliu  PTAL
If this demo meets our requirements?

**Comment by cuisongliu** [comment]
@fanux  Is this certification module written directly to sealos or is it suitable for an independent project?

**Comment by fanux** [comment]
cool!!!

**Comment by fanux** [comment]
> @fanux Is this certification module written directly to sealos or is it suitable for an independent project?

yes

**Comment by cuisongliu** [comment]
1. github.com/wonderivan/logger  => github.com/labring/sealos/pkg/utils/logger
2. k8s client using NewKubernetesClient in client-go pkg
3. idempotency.go Achieved createRoleBinding

**Comment by fanux** [comment]
1. Provide casdoor kubernetes yaml (all in one on openebs).
2. Blog or document about how to integrate casdoor and sealos.
3. Using cloud.sealos.io as the login home page, change login logo.

File: pkg/auth/conf/desktop.yaml.template [hunk 1]
```diff
@@ -0,0 +1,12 @@
+port: 8080
```
Comment (fanux) [inline pkg/auth/conf/desktop.yaml.template:1]: You can provide a config example in docs.

File: go.mod [hunk 2]
```diff
@@ -34,11 +36,13 @@ require (
 	github.com/sirupsen/logrus v1.8.1
 	github.com/spf13/cobra v1.4.0
 	github.com/spf13/pflag v1.0.5
+	github.com/wonderivan/logger v1.0.0
```
Comment (fanux) [inline go.mod:39]: use sealos/pkg/utils/logger

File: pkg/auth/server.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+func Serve() {
+	conf.GlobalConfig = conf.InitConfig(ConfigPath)
+	logger.Info("Loaded configuration")
+	ssoClient = sso.InitSSO()
+	logger.Info("Init SSO platform successfully")
+	httpServer()
+}
+
+func httpServer() {
+	wsContainer := restful.NewContainer()
```
Comment (fanux) [inline pkg/auth/server.go:46]: wsContainer means WebSocket container~

File: pkg/auth/server_config.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+var clientset *kubernetes.Clientset
+
+func ConfigRegister(auth *restful.WebService)  {
+	auth.Path("/").
+		Consumes("*/*").
+		Produces(restful.MIME_JSON) // you can specify this per route as well
+	//user token http
+	auth.Route(auth.GET("/login").To(handlerLogin))
```
Comment (fanux) [inline pkg/auth/server_config.go:30]: add some discribe for those two handler, /login for callback, and /config for got the user kubeconfig?

File: go.mod (hunk 2 above)
Comment (Abingcbc) [reply go.mod]: done

**Comment by cuisongliu** [comment]
pkg/auth/utilsIs it better to split it? For example  k8s into `pkg/clinet-go/kubernetes/idempotency.go`  cors and pki into `pkg/utils/cors`  and  `pkg/utils/pki` ?

File: pkg/auth/utils/config.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+			Type:  "PRIVATE KEY",
+			Bytes: x509.MarshalPKCS1PrivateKey(key),
+		},
+	)
+
+	tml := x509.Certificate{
+		NotBefore: time.Now(),
+		NotAfter:  time.Now().AddDate(99, 0, 0),
+		// you have to generate a different serial number each execution
+		SerialNumber: big.NewInt(123456),
```
Comment (fanux) [inline pkg/auth/utils/config.go:65]: TODO

**Review by fanux — APPROVED** [review]

**Pull request merged**